src/agent.py

```diff
@@ -53,6 +62,65 @@ def __init__(self, charm: ops.CharmBase, state: State):
         charm.framework.observe(
             charm.on[AGENT_RELATION].relation_departed, self._on_agent_relation_departed
         )
+        # Event hooks for agent-discovery-ingress
+        charm.framework.observe(
+            ingress_observer.ingress.on.ready,
+            self.reconfigure_agent_discovery,
+        )
+        charm.framework.observe(
+            ingress_observer.ingress.on.revoked,
+            self.reconfigure_agent_discovery,
+        )
+
+    @property
+    def agent_discovery_url(self) -> str:
+        """Return the external hostname to be passed to agents via the integration.
+
+        If we do not have an ingress, then use the pod ip as hostname.
+        The reason to prefer this over the pod name (which is the actual
+        hostname visible from the pod) or a K8s service, is that those
+        are routable virtually exclusively inside the cluster (as they rely)
+        on the cluster's DNS service, while the ip address is _sometimes_
+        routable from the outside, e.g., when deploying on MicroK8s on Linux.
+
+        Returns:
+            The charm's agent discovery url.
+        """
+        # Check if an ingress URL is available
+        try:
+            if ingress_url := self.ingress_observer.ingress.url:
+                return ingress_url
+        except ops.ModelError as exc:
+            # We only log the error here as we can fallback to using pod IP
+            # if ingress is not available
+            logger.error(
+                "Failed obtaining agent discovery url: %s, is the charm shutting down?", exc
+            )
```

Comment:
It is unusual to have an exception be raised on accessing a property? Perhaps getting the ingress URL should be a function? It also isn't clear why it is an `ops` error please?

Reply:
I didn't make much modification as this was taken from the [prometheus charm](https://github.com/canonical/prometheus-k8s-operator/blob/73cbee1c6ae5150d81bdef0682e67057ffbc6f9a/src/charm.py#L372), as far as I know, `ingress_observer.ingress.url` is a method with `@property` decorator, it also [fetch data from the model's relation data](https://github.com/canonical/traefik-k8s-operator/blob/cb25b7e4fb1f9d20c5a30fa1e8168a16812b1e7e/lib/charms/traefik_k8s/v2/ingress.py#L814) and it can raise `ops.ModelError` as a result

Reply:
I'll update the doc here to add more context

Reply:
Perhaps it also needs to be changed on the prometheus charm - what do you think about the feedback provided above?

Reply:
This implementation is also used in other projects as well and not just the prometheus charm I think. The decision to implementation fetching the ingress URL as a property comes from the traefik charm lib which is used by a lot of charms, so I think this change would be quite significant.

The ModelError raised from this is also something implemented by the traefik lib so I think a comment explaining why we catch that error here would be sufficient in the context of this PR.